go/store/types/value_store_test.go

```diff
@@ -209,6 +212,177 @@ func TestGC(t *testing.T) {
 	assert.Nil(v2)
 }
 
+func TestGCStateDetails(t *testing.T) {
+	// In the absence of concurrency, these tests call
+	// |waitForNoGC| without gcMu held.
+	t.Run("StartsNoGC", func(t *testing.T) {
+		vs := newTestValueStore()
+		vs.waitForNoGC()
+	})
+	t.Run("NewGenBeforeOldGenPanics", func(t *testing.T) {
+		vs := newTestValueStore()
+		assert.Panics(t, func() {
+			vs.transitionToNewGenGC()
+		})
+	})
+	t.Run("FinalizingBeforeNewGenPanics", func(t *testing.T) {
+		vs := newTestValueStore()
+		assert.Panics(t, func() {
+			vs.transitionToFinalizingGC()
+		})
+		vs = newTestValueStore()
+		vs.transitionToOldGenGC()
+		assert.Panics(t, func() {
+			vs.transitionToFinalizingGC()
+		})
+	})
+	t.Run("NoGCAlwaysPossible", func(t *testing.T) {
+		vs := newTestValueStore()
+		vs.transitionToNoGC()
+		vs.waitForNoGC()
+		vs.transitionToOldGenGC()
+		vs.transitionToNoGC()
+		vs.waitForNoGC()
+		vs.transitionToOldGenGC()
+		vs.transitionToNewGenGC()
+		vs.transitionToNoGC()
+		vs.waitForNoGC()
+		vs.transitionToOldGenGC()
+		vs.transitionToNewGenGC()
+		vs.transitionToFinalizingGC()
+		vs.transitionToNoGC()
+		vs.waitForNoGC()
+	})
+	t.Run("transitionToOldGenGC_Concurrent", func(t *testing.T) {
+		vs := newTestValueStore()
+		var wg sync.WaitGroup
+		const numThreads = 16
+		wg.Add(numThreads)
+		var running atomic.Int32
+		for i := 0; i < numThreads; i++ {
+			go func() {
+				defer wg.Done()
+				runtime.Gosched()
```

Comment:
First encounter with this, deeply cursed

Reply:
Might add a note somewhere in this file about what this does (forces another goroutine to start executing in this thread)